src_js/components/sidebar/TableOfContents.tsx

```diff
@@ -0,0 +1,126 @@
+import { h, Fragment } from 'preact';
+import { useState, useEffect } from 'preact/hooks';
+import unflattenHeadings, { HeadingsSectionType } from './unflattenHeadings';
+
+export type PropsType = {
+  contentNodeSelector: string;
+  isSmallScreen: boolean;
+  sidebarShown: boolean;
+  settingsShown: boolean;
+  topbarHeight: number;
+  onToggleSidebar: () => void;
+  onToggleSettings: () => void;
+};
+
+export default function TableOfContents(props: PropsType) {
+  // We don't actually care about the window.scrollY state, we can get that
+  // from the window object on demand. We use a state-setter so that we can
+  // re-render the component.
+  const [_, setWindowScrollDistance] = useState(window.scrollY || 0);
+
+  // When the user scrolls, rerender the component.
+  useEffect(() => {
+    const scrollHandler = () => setWindowScrollDistance(window.scrollY);
+    window.addEventListener('scroll', scrollHandler);
+    return () => {
+      window.removeEventListener('scroll', scrollHandler);
+    };
+  }, []);
+
+  // When a TOC item is clicked, close the Settings. Also close the Sidebar on
+  // small screens.
+  useEffect(() => {
+    const hash_listener = () => {
+      if (props.isSmallScreen && props.sidebarShown) {
+        props.onToggleSidebar();
+      }
+      if (props.settingsShown) {
+        props.onToggleSettings();
+      }
+    };
+
+    window.addEventListener('hashchange', hash_listener, false);
```

Comment:
One issue with listening to `hashchange` is that if the user clicks the same sidebar item consecutively, subsequent clicks do not trigger this event. (So on mobile, if a user clicks a section, scrolls, then opens the sidebar to click the same section, the sidebar won't close.)

I think this is an acceptable bug for now, but it should be documented by opening an issue.